# Formateo de cadenas

El formateo de cadenas nos permite inyectar intems en una cadena, sin tener que ir concatenándolos. Veamos unapequeña comparativa:

    jugador = 'Juan'
    puntos = 33
    
    'Esta noche, '+jugador+' anotó '+str(puntos)+' puntos.'  # concatenación
    
    f'Esta noche, {jugador} anotó {puntos} punts.'          # formateo de cadenas


Hay tres formas de llevar a cabo el formateo de cadenas.
* El método antiguo que usa el carácter porcentaje `%`.
* Un método mejorado que usa el método de cadenas  `.format()`.
* El método más nuevoThe, introducido en Python 3.6, que usa cadenas literales formateadas, llamadas *f-strings*.

Como podemos encontrarnos las tres versiones al leer el código de otro programador, veremos ejemplos de los tres.

## Formateando con %
Puedes usar <code>%s</code> para inyectar cadenas en tus sentencias de impresión. El carácter `%` hace referencia a un "operador de formateo de cadenas".

In [1]:
print("Voy a inyectar %s aquí." %'algo')

Voy a inyectar algo aquí.


Puedes pasar varios objetos poniendolos dentro de una tupla después del operador `%`.

In [2]:
print("Voy a inyectar %s texto aquí, y %s texto aquí." %('algún','más'))

Voy a inyectar algún texto aquí, y más texto aquí.


También podemos pasar nombres de variables:

In [3]:
x, y = 'algún', 'más'
print("Voy a inyectar %s texto aquí, y %s texto aquí."%(x,y))

Voy a inyectar algún texto aquí, y más texto aquí.


### Métodos de conversión de formato.
Hay que mencionar que los dos métodos <code>%s</code> y <code>%r</code> convierten cualquier objeto de python en cadena usando dos métodos separados: `str()` y `repr()`. Aprenderemos más acerca de estas funciones más adelante, pero hay que destacar que  `%r` y `repr()` entregan una *representación en cadena* del objeto, incluyendo comillas y cualquier carácter de escape.

In [4]:
print('Dijo que su nombre era %s.' %'Fred')
print('Dijo que su nombre era %r.' %'Fred')

Dijo que su nombre era Fred.
Dijo que su nombre era 'Fred'.


Otro ejemplo, `\t` inserta una tabulación en una cadena.

In [5]:
print('Una vez capturé un pescado %s.' %'asi \tde grande')
print('Una vez capturé un pescado %r.' %'asi \tde grande')

Una vez capturé un pescado asi 	de grande.
Una vez capturé un pescado 'asi \tde grande'.


El operador `%s` convierte lo que ve en una cadena, incluyendo enteros y decimales. El operador `%d` convierte números a enteros, sin redondeo. Ejemplo:

In [6]:
print('He escrito %s programas hoy.' %3.75)
print('He escrito %d programas hoy.' %3.75)   

He escrito 3.75 programas hoy.
He escrito 3 programas hoy.


### Padding y Precisión de números en coma flotante
Números en coma flotante usan el formato <code>%5.2f</code>. Aquí, <code>5</code> sería el menor número de caractéres que la cadena debe contener; estos deberían ser rellenados con espacios en blanco si el número completo no contiene tantos dígitos. Además, <code>.2f</code> significa cuántos números tras el punto decimal se mostrarán. Veamos algunos ejemplos:

In [8]:
print('Numeros en coma flotante: %5.2f' %(13.144))

Numeros en coma flotante: 13.14


In [9]:
print('Numeros en coma flotante: %1.0f' %(13.144))

Numeros en coma flotante: 13


In [10]:
print('Numeros en coma flotante: %1.5f' %(13.144))

Numeros en coma flotante: 13.14400


In [11]:
print('Numeros en coma flotante: %10.2f' %(13.144))

Numeros en coma flotante:      13.14


In [12]:
print('Numeros en coma flotante: %25.2f' %(13.144))

Numeros en coma flotante:                     13.14


Para más información podéis visitar https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Formateo múltiple
No hay nada que impida usar más de un método de conversión en la misma declaración:

In [13]:
print('Primero: %s, Secundo: %5.2f, Tercero: %r' %('hola!',3.1415,'adios!'))

Primero: hola!, Secundo:  3.14, Tercero: 'adios!'


## Formateando con el método `.format()` 
Una forma mejor de formatear objetos en tus cadenas en tus sentencias de impresión es usar el método `.format()`. La sintáxis es:

    'Cadena aquí {} y aquí {}'.format('algo1','algo2')
    
Por ejemplo:

In [14]:
print('Esto es una cadena con una  {}'.format('insercción'))

Esto es una cadena con una  insercción


### El método  .format() tiene varias ventajas sobre el método %s :

#### 1. Los objetos insertados se pueden llamar por su índice:

In [17]:
print('El {2} {1} {0}'.format('marrón','zorro','rápido'))

El rápido zorro marrón


#### 2. A los objetos insertados le les puede asignar palabras clave:

In [19]:
print('Primer objeto: {a}, Segundo objeto: {b}, Tercer objeto: {c}'.format(a=1,b='Dos',c=12.3))

Primer objeto: 1, Segundo objeto: Dos, Tercer objeto: 12.3


#### 3. Los objetos insertados pueden ser reutilizados, evitando la duplicación:

In [20]:
print('Un %s ahorrado es un %s ganado.' %('euro','euro'))
# vs.
print('Un {p} ahorrado es un {p} ganado.'.format(p='euro'))

Un euro ahorrado es un euro ganado.
Un euro ahorrado es un euro ganado.


### Alineación, espaciado y precisión con `.format()`
En las llaves puedes asignar longitud del campo, alineación izquierda/derecha, parámetros de redondeo y más

In [21]:
print('{0:8} | {1:9}'.format('Fruta', 'Cantidad'))
print('{0:8} | {1:9}'.format('Manzanas', 3.))
print('{0:8} | {1:9}'.format('Naranjas', 10))

Fruta    | Cantidad 
Manzanas |       3.0
Naranjas |        10


Por defecto, `.format()` alinea el texto a la izquierda, y los números a la derecha. Podemos pasar un argumento adicional `<`,`^`, o `>` para definir una alineación izquierda, centrada o derecha:

In [22]:
print('{0:<8} | {1:^8} | {2:>8}'.format('izquireda','centro','derecha'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

izquireda |  centro  |  derecha
11       |    22    |       33


Puedes preceder el operador de alineación con un carácter de relleno

In [23]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('izquierda','cenro','derecha'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

izquierda | -cenro-- | .derecha
11====== | ---22--- | ......33


Los campor ancho y precisión se manejan de forma similar al método que hemos visto. Las siguientes sentencias de impresión son equivalentes:

In [24]:
print('Este es mi diez-decimales, dos-decimales número:%10.2f' %13.579)
print('Este es mi diez-decimales, dos-decimales número:{0:10.2f}'.format(13.579))

Este es mi diez-decimales, dos-decimales número:     13.58
Este es mi diez-decimales, dos-decimales número:     13.58


Fijate que hay 5 espacios tras los dos puntos, y 5 caracteres contando  13.58, hasta un total de 10 caracteres.

Para más información sore este método, puedes visitar https://docs.python.org/3/library/string.html#formatstrings

## Cadenas literales formateadas (f-strings)

El método introducido desde la versión 3.6 de Python, f-strings, ofrece bastantes beneficios sobre el método antiguo `.format()`.

In [26]:
nombre = 'Fred'

print(f"Él dijo que su nombre es {nombre}.")

Él dijo que su nombre es Fred.


Pasamos `!r` para obtener la representación de la cadena:

In [27]:
print(f"Él dijo que su nombre es {nombre!r}")

Él dijo que su nombre es 'Fred'


#### Más de formateo de decimales `"resultado: {valor:{ancho}.{precisión}}"`

Mientras con el método `.format()` verías `{valor:10.4f}`, con el método f-strings verás `{valor:{10}.{6}}`


In [28]:
num = 23.45678
print("Mis 10 caracteres, 4 números decimales es:{0:10.4f}".format(num))
print(f"Mis 10 caracteres, 4 números decimales es:{num:{10}.{6}}")

Mis 10 caracteres, 4 números decimales es:   23.4568
Mis 10 caracteres, 4 números decimales es:   23.4568


Nótese que con f-strings, *precisión* se refiere al total de números de dígitos, no solo los que hay tras el decimal. Esto se ajusta más a la notación científica y el análisis estadísticoThis.

In [29]:
num = 23.45
print("Mis 10 caracteres, 4 números decimales es:{0:10.4f}".format(num))
print(f"Mis 10 caracteres, 4 números decimales es:{num:{10}.{6}}")

Mis 10 caracteres, 4 números decimales es:   23.4500
Mis 10 caracteres, 4 números decimales es:     23.45


También podemos usar `.format()` dentro de una f-string:

In [30]:
num = 23.45
print("Mis 10 caracteres, 4 números decimales es:{0:10.4f}".format(num))
print(f"Mis 10 caracteres, 4 números decimales es:{num:10.4f}")

Mis 10 caracteres, 4 números decimales es:   23.4500
Mis 10 caracteres, 4 números decimales es:   23.4500


Para más infirmación puedes visitar https://docs.python.org/3/reference/lexical_analysis.html#f-strings

That is the basics of string formatting!